src/frontend/src/expr/mod.rs

```diff
@@ -176,18 +183,60 @@ impl ExprImpl {
         matches!(self, ExprImpl::Literal(literal) if literal.return_type() == DataType::Varchar)
     }
 
+    pub fn cast_implicit_or_shrink(self, target: DataType) -> Result<CastResult> {
+        let implicit_error = match self.clone().cast_implicit(target.clone()) {
+            Ok(expr) => return Ok(CastResult::Success(expr)),
+            Err(err) => err,
+        };
+        match self.cast_shrink(target) {
+            Ok(CastResult::Failed) | Err(_) => Err(implicit_error.into()),
+            ok => ok,
+        }
+    }
+
+    pub fn cast_shrink(self, target: DataType) -> Result<CastResult> {
+        match (self.return_type(), &target) {
+            (DataType::Int64, DataType::Int32)
+            | (DataType::Int64, DataType::Int16)
+            | (DataType::Int32, DataType::Int16) => self.cast_shrink_for_int(target),
+            _ => Ok(CastResult::Failed),
+        }
+    }
+
+    fn cast_shrink_for_int(self, target: DataType) -> Result<CastResult> {
+        let (upper_bound, lowwer_bound) = match (self.return_type(), &target) {
+            (DataType::Int64, DataType::Int32) => (i32::MAX as i64, i32::MIN as i64),
+            (DataType::Int64, DataType::Int16) | (DataType::Int32, DataType::Int16) => {
+                (i16::MAX as i64, i16::MIN as i64)
+            }
+            _ => return Ok(CastResult::Failed),
+        };
+        if let Some(scalar) = self.eval_row_const()? {
+            let value = scalar.as_integral();
+            if value > upper_bound {
+                Ok(CastResult::OutUpperBound)
+            } else if value < lowwer_bound {
+                Ok(CastResult::OutLowerBound)
+            } else {
+                Ok(CastResult::Success(self.cast_explicit(target)?))
+            }
+        } else {
+            Ok(CastResult::Failed)
+        }
+    }
+
     /// Shorthand to create cast expr to `target` type in implicit context.
-    pub fn cast_implicit(self, target: DataType) -> Result<ExprImpl> {
+    pub fn cast_implicit(self, target: DataType) -> std::result::Result<ExprImpl, ErrorCode> {
```

Comment:
Hmm, why is it necessary to change the error type?

Reply:
For RwError, it will capture the backtrace so the cost of constructing the RwError is high. There is a known issue #6131. 

In cast_implicit_or_shrink, We will store Error return form cast_implicit before call cast_shrink, so if cast_implicit return a RwError, the performance is unacceptable.